### DISCUSSION TITLE
[NRT] Allow `MemberNotNullAttribute`/`MemberNotNullWhen` on parameters

### DISCUSSION THREAD

**Discussion by gurustron** [dotnet/csharplang#6915] [Language Ideas] [answered]
Sorry, was not able to find matching issue/discussion (though sure there is one). 

Consider following scenario:

```csharp
public class A
{
    public string? Prop { get; set; }
}

public static class AExts
{
    public static bool Validate([NotNullWhen(returnValue:true)]this A? a) => 
        a is not null 
        && string.IsNullOrEmpty(a.Prop);
} 
```

Then obviously we will get the following behavior:

```csharp
public static bool Do(A? a)
{
    if (!a.Validate())
    {
        return false;
    }

    Console.WriteLine(a.ToString()); // no warning
    Console.WriteLine(a.Prop.ToUpper()); // warning
    return true;
}
```

It would be nice to have a way to inform the compiler that `Prop` was validated too. Something like:

```csharp
public static bool Validate(
    [NotNullWhen(returnValue:true)]
    [MemberNotNullWhen(returnValue:true, nameof(a.Prop))]
    this A? a) => 
    a is not null 
    && string.IsNullOrEmpty(a.Prop);
```

Potential considerations:

- Handle paths (i.e. `[MemberNotNullWhen(returnValue:true, nameof(a.Prop.NestedProp.AnotherNestedProp))]`)
- Allow passing multiple properties in one attribute
- Consider for other attributes like `NotNullIfNotNull`

**Comment by RikkiGibson** [comment, marked as answer]
#4563
